src/main/java/com/avail/optimizer/L2Optimizer.java

```diff
@@ -440,20 +580,60 @@ void postponeConditionallyUsedValues ()
 							edge.alwaysLiveInRegisters.addAll(
 								newInstruction.sourceRegisters());
 						}
-						block.instructions().remove(instruction);
-						instruction.justRemoved();
+						// Erase its writes in every path at and after the edges
+						// that we *didn't* move it to.  We don't have to worry
+						// about hitting a block that does care about the value,
+						// because we already know it was never live-in anywhere
+						// after this edge.
+						reachableTargetEdges.forEach(edge ->
+						{
+							if (!edgesToMoveThrough.contains(edge))
+							{
+								edge.forgetRegistersInManifestsRecursively(
+									writtenSet);
+							}
+						});
+
+						// We moved the instruction *just* out of the block, so
+						// when we move prior instructions, they'll still end up
+						// before this one.  Therefore, don't track its reads.
+						// Similarly, its writes have also moved past where any
+						// new instructions might get moved.
 					}
 					else
 					{
 						// The instruction stayed where it was.  Record the
-						// registers that it consumed, to pin any prior
-						// instructions that provide those values, since they
-						// can't be moved out of the block.
+						// registers or semantic values that it consumed, to pin
+						// any prior instructions that provide those values,
+						// since they also can't be moved out of the block.
 						registersConsumedLaterInBlock.addAll(
 							instruction.sourceRegisters());
+						instruction.readOperands().stream()
+							.map(L2ReadOperand::semanticValue)
+							.collect(
+								toCollection(
+									() -> semanticValuesConsumedLaterInBlock));
+						// It might read values that prior instructions write,
+						// so we should prevent those instructions from moving.
+						dependentReadsLaterInBlock |=
+							operation.readsHiddenVariablesMask;
+						// It might overwrite hidden values needed by prior
+						// instructions, so prevent those readers from moving.
+						allWritesLaterInBlock |=
+							operation.writesHiddenVariablesMask;
+
+						// Also record any mid-block exit edges, so that prior
+						// instructions can decide whether to replicate there as
+						// well.  At the moment, these don't exist.
```

Comment:
"Mid-block exit edges"? Isn't this conceptually a violation of basic block structure? Maybe I'm misunderstanding the sentiment here.

Reply:
They existed for a while for L2_SAVE_ALL_AND_PC_TO_INT, but they weren't worth the extra complexity.  Updated the vestigial comments.